services/graph/pkg/unifiedrole/roles.go

```diff
@@ -536,35 +536,58 @@ func GetLegacyRoleName(role libregraph.UnifiedRoleDefinition) string {
 // descending - false - sorts the roles from least to most permissions
 // descending - true - sorts the roles from most to least permissions
 func weightRoles(roleSet []*libregraph.UnifiedRoleDefinition, constraints string, descending bool) []*libregraph.UnifiedRoleDefinition {
-	slices.SortFunc(roleSet, func(i, j *libregraph.UnifiedRoleDefinition) int {
-		var ia []string
-		for _, rp := range i.GetRolePermissions() {
-			if rp.GetCondition() == constraints {
-				ia = append(ia, rp.GetAllowedResourceActions()...)
-			}
+	specialIDs := []string{
+		UnifiedRoleSecureViewerID,
+		UnifiedRoleDeniedID,
+	}
+
+	specialSet := make(map[string]struct{}, len(specialIDs))
+	for _, id := range specialIDs {
+		specialSet[id] = struct{}{}
+	}
+
+	normalRoles := make([]*libregraph.UnifiedRoleDefinition, 0, len(roleSet))
+	specialRoles := make([]*libregraph.UnifiedRoleDefinition, 0, len(roleSet))
```

Comment:
```suggestion
	specialRoles := make([]*libregraph.UnifiedRoleDefinition, 0, len(specialIDs))
```

... because you can never exceed the length of the specialIDs, so this is a good upper bound.